types.go

```diff
@@ -0,0 +1,296 @@
+	RemoveServer(id raft.ServerID, prevIndex uint64, timeout time.Duration) raft.IndexFuture
+	Stats() map[string]string
+	LeadershipTransferToServer(id raft.ServerID, address raft.ServerAddress) raft.Future
+}
+
+type Delegate interface {
+	// AutopilotConfig is used to retrieve the latest configuration from the delegate
+	AutopilotConfig() *Config
+
+	// NotifyState will be called when the autopilot state is updated. The application may choose to emit metrics
+	// or perform other actions based on this information.
+	NotifyState(*State)
+
+	// FetchServerStats will be called to request the application fetch the ServerStats out of band. Usually this
+	// will require an RPC to each server.
+	FetchServerStats(context.Context, map[raft.ServerID]*Server) map[raft.ServerID]*ServerStats
+
+	// KnownServers fetchs the list of servers as known to the application
+	KnownServers() map[raft.ServerID]*Server
+
+	// RemoveFailedServer notifies the application to forcefully remove the server in the failed state
+	RemoveFailedServer(*Server) error
+}
+
+type RaftChanges struct {
+	Promotions []raft.ServerID
+	Demotions  []raft.ServerID
+	Leader     raft.ServerID
+}
+
+type FailedServers struct {
+	// StaleNonVoters are the ids of those server in the raft configuration as non-voters
+	// that are not present in the delegates view of what servers should be available
+	StaleNonVoters []raft.ServerID
+	// StaleVoters are the ids of those servers in the raft configuration as voters that
+	// are not present in the delegates view of what servers should be available
+	StaleVoters []raft.ServerID
+
+	// FailedNonVoters are the servers without voting rights in the cluster that the
+	// delegate has indicated are in a failed state
+	FailedNonVoters []*Server
+	// FailedVoters are the servers without voting rights in the cluster that the
+	// delegate has indicated are in a failed state
+	FailedVoters []*Server
+}
+
+// Promoter is an interface to provide promotion/demotion algorithms to the core autopilot type.
+// The BasicPromoter satisfies this interface and will promote any stable servers but other
+// algorithms could be implemented. The implementation of these methods shouldn't "block".
+// While they are synchronous autopilot expects the algorithms to not make any network
+// or other requests which way cause an indefinite amount of waiting to occur.
+//
+// Note that all parameters passed to these functions should be considered read-only and
+// their modification could result in undefined behavior of the core autopilot routines
+// including potential crashes.
+type Promoter interface {
+	// GetServerExt returns some object that should be stored in the Ext field of the Server
+	// This value will not be used by the code in this repo but may be used by the other
+	// Promoter methods and the application utilizing autopilot. If the value returned is
+	// nil the extended state will not be updated.
+	GetServerExt(*Config, *ServerState) interface{}
```

Comment:
I guess another way this could work would be for the `Promoter` to maintain a map of `Server` instances to whatever metadata it needs, that way when another one its methods is called, it can use that map instead of storing the metdata on the `Server` as an opaque field. I'm not sure how well that approach would work here as I haven't had a chance to look at how `Ext` would be used.

Reply:
We want to expose the promoter information through the main autopilot APIs. If we had them track the state themselves we would have to stitch things back together later on. Which I guess would be entirely possible. I will have to think on this a bit more.

Also the plan was to version this library as v0.1.0 initially so we can break the interface if we want to in the future. Not that we don't want to try and get it right now of course but we could try it out and see how things go for Nomad integrating the library too.

Reply:
I guess in the case where the promoter keeps state we we would reduce the interface to just three methods

```
type Promoter interface{
   UpdateState(*Config, *State) (map[raft.ServerID]NodeType, error)
   CalculateRaftChanges(*Config, *State) error
   FilterFailedServerRemovals(*Config, *State, *FailedServers) *FailedServers
}
```

Reply:
Actually there would be consistency issues with keeping state in the Promoter similar to the Config.

When the application goes to read the current state it would now have to pull from autopilot and the promoter. Whether the two are supposed to go together would be timing dependent as the promoter could have its state updated at the same time as we are reading the autopilot state.